app/controllers/spree/afterpay_controller.rb

```diff
@@ -0,0 +1,78 @@
+module Spree
+  class AfterpayController < Spree::BaseController
+
+    before_action :load_order
+    before_action :validate_token, only: [:cancel, :success]
+    before_action :validate_success_status, only: [:success]
+    before_action :reset_completed_at_field_of_order, only: [:cancel, :success]
+    before_action :invalidate_afterpay_payment, only: [:cancel]
+
+    def success
+      complete_order(Spree.t(:order_processed_successfully))
+    end
+
+    def cancel
+      if params[:status] && params[:status] == 'CANCELLED'
+        flash[:error] = Spree.t(:payment_cancelled, scope: [:afterpay])
+      else
+        flash[:error] = Spree.t(:payment_failed, scope: [:afterpay])
+      end
+      redirect_to checkout_state_path(@order.state)
+    end
+
+    private
+
+    def validate_token
+      sanitized_token = params[:orderToken].try(:split, '?').try(:first)
+      unless sanitized_token == @order.payments&.afterpay.last&.source&.transaction_id
+        redirect_to(spree.cart_path) && return
+      end
+    end
+
+    def validate_success_status
+      unless params[:status] && params[:status] == "SUCCESS"
+        flash[:error] = Spree.t(:payment_failed, scope: [:afterpay])
+        redirect_to checkout_state_path(@order.state)
+      end
+    end
+
+    def reset_completed_at_field_of_order
+      if @order.checkout_afterpay_payments.present?
+        @order.update_columns(completed_at: nil)
+      end
+    end
+
+    def completion_route
+      spree.order_path(@order)
+    end
+
+    def load_order
+      @order = Spree::Order.includes(:adjustments, line_items: [variant: [:images, :option_values, :product]]).find_by(number: params[:id])
+
+      unless @order.present? && @order.payment_total < @order.total
+        redirect_to(spree.cart_path) && return
+      end
+    end
+
+    def invalidate_afterpay_payment
+      @order.checkout_afterpay_payments.map(&:invalidate!)
+    end
+
+    def complete_order(flash_message)
+      unless @order.next
```

Comment:
so, first you are charging the customer and then you are checking if the order state can be moved... you should do the opposite instead... and if the charge is failed simply revert the state